### DISCUSSION TITLE
Propose: Add "where" support for enum in generic type defintions

### DISCUSSION THREAD

**Discussion by The-Futurist** [dotnet/csharplang#329] [General]
Currently there's no way to restrict a generic type argument to be an enum. One "workaround" is to restrict the type to a struct perhaps adding IComparable etc to the "where" clause.

So this is a proposal to add this feature to C#.

This was inspired by this related issue:

https://github.com/dotnet/roslyn/issues/18126

This has come up numerous time in certain kinds of designs, it isn't a frequent issue but does arise and does seem like a gap in the language.

Thx

**Comment by HaloFour** [comment]
Dupe of #104.

**Comment by temporaryfile** [comment]
#104 just disappeared, so...

Generic constraints make extremely good sense with enum inheritance. Example from the future:

```
public enum WindowMessage
{
    WM_NULL = 0,
}

public enum ListViewMessage : WindowMessage
{
    LVM_FIRST = 0x1000,
    LVM_EDITLABELW = (LVM_FIRST + 118),
}

public static IntPtr SendMessage<TMessage>(
    HWND hwnd,
    TMessage Msg = WindowMessage.WM_NULL,
    UIntPtr wParam = default(UIntPtr),
    IntPtr lParam = default(IntPtr))
    where TMessage : WindowMessage
{
    // [....]
}
```

Wait till I show you the example with NT kernel object access flags.
